compiler/rustc_trait_selection/src/solve/fulfill.rs

```diff
@@ -258,5 +265,135 @@ fn fulfillment_error_for_stalled<'tcx>(
         }
     });
 
-    FulfillmentError { obligation: obligation.clone(), code, root_obligation: obligation }
+    FulfillmentError {
+        obligation: find_best_leaf_obligation(infcx, &obligation),
+        code,
+        root_obligation: obligation,
+    }
+}
+
+struct BestObligation<'tcx> {
+    obligation: PredicateObligation<'tcx>,
+}
+
+impl<'tcx> BestObligation<'tcx> {
+    fn with_derived_obligation(
+        &mut self,
+        derive_obligation: impl FnOnce(&mut Self) -> PredicateObligation<'tcx>,
+        and_then: impl FnOnce(&mut Self) -> <Self as ProofTreeVisitor<'tcx>>::Result,
+    ) -> <Self as ProofTreeVisitor<'tcx>>::Result {
+        let derived_obligation = derive_obligation(self);
+        let old_obligation = std::mem::replace(&mut self.obligation, derived_obligation);
+        let res = and_then(self);
+        self.obligation = old_obligation;
+        res
+    }
+}
+
+impl<'tcx> ProofTreeVisitor<'tcx> for BestObligation<'tcx> {
+    type Result = ControlFlow<PredicateObligation<'tcx>>;
+
+    fn span(&self) -> rustc_span::Span {
+        self.obligation.cause.span
+    }
+
+    fn visit_goal(&mut self, goal: &super::inspect::InspectGoal<'_, 'tcx>) -> Self::Result {
+        let candidates = goal.candidates();
+        // FIXME: Throw out candidates that have no failing WC and >1 failing misc goal.
+
+        // HACK:
+        if self.obligation.recursion_depth > 3 {
+            return ControlFlow::Break(self.obligation.clone());
+        }
```

Comment:
lmao. This could be solved some other way, like giving a `nested_goal.super_visit(self)` which can be used below instead of `self.visit_goal(&nested_goal)`, and which checks the goal's depth against the config.